### DISCUSSION TITLE
Asynchronously creating a score token

### DISCUSSION THREAD

**Discussion by GiantNuker** [ppy/osu#15353] [Q&A] [answered]
(not sure if this is the right discussion category or if I should've just made a PR)

Currently, a score token is requested from the api whenever a `SubmittingPlayer` is created,in `handleTokenRetrieval`. (`SubmittingPlayer` line 41)
The problem is that `handleTokenRetrieval` is called from `LoadAsyncComplete`, which stops the map from loading until it finishes.

My solution was to wrap `handleTokenRetrieval` in `Task.Factory.StartNew`.
This solution *does* work as intended, but I'm fairly sure it should go into a network thread pool somewhere, which i cant seem to find.

Basically, whats the correct way to call `handleTokenRetrieval` asynchronously? (Or why its a terrible idea/shouldn't be implemented?)

Edit: the reason i notice this at all is that my internet is terrible and it adds a fairly large (sometimes upwards of 10 seconds) delay to map loading

**Comment by peppy** [comment, marked as answer]
For now this is intentional, as we don't want players playing under the false premise that their play is going to be ranked when it isn't.

Also going forward there will be other reasons why the token is required before starting gameplay.

Can you describe how your internet can be so bad it takes 10 seconds to do a very small web request? This is going to cause large issues for you game-wide, not just with score token retrieval.

**Reply by GiantNuker** [reply]
Maybe at least having it time out after a few seconds? Like 5-10ish? It just keeps going until it fails atm?

> Can you describe how your internet can be so bad it takes 10 seconds to do a very small web request? 

Basically the internet for our whole house is a mobile hotspot off a cell booster 15 miles away from the nearest cell tower on a deprioritized data plan.
Most of the day if i run something like `ping google.com`  it will be over 500ms, with packet loss.

> This is going to cause large issues for you game-wide, not just with score token retrieval.

If its about the same as stable thats fine (sometimes multiplayer works, score submission works almost all the time)

I'm assuming this is for more server sided anticheat (did the player complete this 4 minute map in 0.3 seconds?), in which case it wont.

Edit:
I'm assuming an option to just not submit scores wont be a thing, and neither will running in debug mode and connecting to the live server

**Reply by bdach** [reply]
> I'm assuming an option to just not submit scores wont be a thing

Well this is an option if you log off. As is they will not submit ever, though, since they won't be associated with your user.

**Reply by GiantNuker** [reply]
oh true, completely forgot that
logging out when it will time out 100% of the time anyway is a good idea

one other idea, idk how viable this one is:
Could you reuse the original score token when retrying a map? That would make it much less annoying.

and lastly is score resubmission (like how stable retries for at least an hour) planned or no?
